nexus/mgs-updates/src/driver_update.rs

```diff
@@ -0,0 +1,576 @@
+    SpUpdateFailed(Uuid, String),
+    #[error("SP not knowing about our update attempt")]
+    SpUpdateLost,
+    #[error(
+        "gave up after {}ms waiting for update {0} to finish",
+        .1.as_millis())
+    ]
+    StuckUpdating(Uuid, Duration),
+    #[error("failed to abort in-progress SP update")]
+    SpUpdateAbortFailed(#[from] AbortError),
+    #[error("SP reports that reset failed: {0:?}")]
+    SpResetFailed(String),
+
+    #[error("failed waiting for artifact delivery")]
+    DeliveryWaitError(#[from] DeliveryWaitError),
+    #[error("error communicating with MGS")]
+    UpdateStartError(GatewayClientError),
+    #[error(
+        "timed out after {}ms waiting for update to finish",
+        .0.as_millis()
+    )]
+    ResetTimeoutError(Duration),
+    #[error("waiting for update to finish")]
+    WaitError(PrecheckError),
+}
+
+/// Makes one complete attempt to apply the specified software update to an SP
+/// component.
+///
+/// `sp_update` specifies the target component, expected previous version,
+/// expected new version, etc.
+///
+/// This operation is made up of multiple steps that have to happen in order:
+/// upload the artifact to MGS, wait for the SP to finish writing it,
+/// potentially change the active slot, reset the device, etc.  At any point,
+/// it's possible that a different Nexus instance decides we're out to lunch and
+/// aborts the process and potentially even starts its own.  While this function
+/// will poll forever as long as it looks like things are working, it will
+/// return early if not (e.g., if we find that somebody else has aborted our
+/// update or if there's been no progress for too long).
+///
+/// On success, `UpdateCompletedHow` describes what was needed to complete the
+/// update.
+pub(crate) async fn apply_update(
+    artifacts: Arc<ArtifactCache>,
+    sp_update: &SpComponentUpdate,
+    update_helper: &(dyn SpComponentUpdateHelper + Send + Sync),
+    mgs_rx: watch::Receiver<AllBackends>,
+    update: &PendingMgsUpdate,
+    status: UpdateAttemptStatusUpdater,
+) -> Result<UpdateCompletedHow, ApplyUpdateError> {
+    // Set up an instance of `MgsClients` to talk to MGS for the duration of
+    // this attempt.  For each call to `try_serially()`, `MgsClients` will try
+    // the request against each MGS client that it has.  That makes it possible
+    // to survive transient failure of MGS as long as one is working.  If all
+    // are offline, these operations will fail.  We rely on the higher-level
+    // operation retry to deal with that.
+    status.update(UpdateAttemptStatus::FetchingArtifact);
+    let log = &sp_update.log;
+    let mut mgs_clients = {
+        let backends = mgs_rx.borrow();
+        if backends.is_empty() {
+            return Err(ApplyUpdateError::NoMgsBackends);
+        }
+        MgsClients::from_clients(backends.iter().map(
+            |(backend_name, backend)| {
+                gateway_client::Client::new(
```

Comment:
I vaguely remember progenitor clients being relatively expensive to construct. Should the qorb pool be handing out already-construct (and cached-in-the-pool) clients instead of just the addresses?

Reply:
I'd like to take this as a follow-up.

@smklein Right now I'm having this thing accept a `watch::Receiver<AllBackends>` (what qorb resolvers provide) and then constructing the clients each time this function is called.  is there some other interface I could be using here to maintain a pool of clients?